Moved ChartUtils drawing methods into CGContext extension

jjatie opened the pull request:
Much nicer call sites
Removed `NSAttributedStringKey` where type inference was sufficient.
Minor tidy of drawText calls in AxisRenderers

It is important to note that this is a breaking change because some of these methods were made public. However, it is extremely unlikely there are direct calls to them so I feel it is safe to merge them into the master. In 4.0.0 I would like to make all of these methods `internal` or at the very least, `public`, as our extension points don't need be at such a low level.

jjatie commented:
@liuxuan30 This looks like a big one, but I didn't change any code, just updated the call sites to reflect the `CGContext` extension instead of `ChartUtils`

liuxuan30 commented:
sure. I will take this. Just need some time

liuxuan30 commented:
jesus, this has 1000+ diff? that gonna take a long time as execuse :)

liuxuan30 commented on Charts.xcodeproj/project.pbxproj:
```diff
@@ -1,1235 +1,1235 @@
 // !$*UTF8*$!
 {
-	archiveVersion = 1;
```
Is this indent stuff intended fix by you @jjatie or just your Xcode auto-fixed it?

jjatie replied:
Yea, I didn't touch any of that...

jjatie replied:
I double checked in BBEdit, and the xcodeproj file is the same. I don't know why those indents happened.

jjatie commented:
1190 of them are all in the xcodeproj file. I don't know why it was changed at all, I'm guessing I just need to pull in recent changes.

jjatie commented:
@liuxuan30 1190 of them are all in the xcodeproj file. I don't know why it was changed at all, I'm guessing I just need to pull in recent changes. If you are able to get this one done today I would be able to start getting to work on some bigger changes for 4.0.0 that will take me a long time.

liuxuan30 commented:
it happens in https://github.com/danielgindi/Charts/pull/3008 as well.
And it now has conflicts. Something must be wrong..

jjatie commented:
I checked line by line this morning, and couldn't find any differences.

liuxuan30 commented:
any idea why it now has conflicts? I will check 4.0 branch later. eating

jjatie commented:
No idea. It's been bugging me

jjatie commented:
Makes no sense to me. I just rebased, and took the pbxproj file from the master and there are STILL conflicts in it. I still don't see any difference other than white space though.

liuxuan30 commented:
I had a theory. Have you recently changed your tab/space stuff? like disabled using space instead tab, etc.

I checkout 4.0 branch, reset --mix of your current commit, and 
Charts.xcodeproj/project.pbxproj is indeed showing diff, however, in my diff tool, it looks like:
![image](https://user-images.githubusercontent.com/4375169/34646712-73075f44-f3aa-11e7-8d30-89e34606d592.png)
And the diff on github, the old one is showing as tab, and thew new one is showing as space. and 1 tab: 4 space. You can try select them and find out this.

liuxuan30 commented on Source/Charts/Charts/PieChartView.swift:
```diff
@@ -397,11 +397,10 @@ open class PieChartView: PieRadarChartViewBase
-                    NSAttributedStringKey.font: NSUIFont.systemFont(ofSize: 12.0),
-                    NSAttributedStringKey.paragraphStyle: paragraphStyle
-                    ], range: NSMakeRange(0, attrString!.length))
+                attrString?.setAttributes([.foregroundColor: NSUIColor.black,
```
shall we use another line?

jjatie replied:
I looked through the framework, and it was very inconsistent for this case. I decided to go with the one I thought was cleanest.

jjatie replied:
To follow the code style of the `{` it should look like this:
```
                attrString?.setAttributes([
                    .foregroundColor: NSUIColor.black,
                    .font: NSUIFont.systemFont(ofSize: 12.0),
                    .paragraphStyle: paragraphStyle
                    ],
                                          range: NSMakeRange(0, attrString!.length))
```

liuxuan30 replied:
sure. NP

liuxuan30 commented on Source/Charts/Renderers/BarChartRenderer.swift:
```diff
@@ -445,12 +445,9 @@ open class BarChartRenderer: BarLineScatterCandleBubbleRenderer
-                                y: py,
-                                size: icon.size)
+                            context.drawImage(icon,
+                                              atCenter: CGPoint(x: px, y: py),
```
will it be possible it's not center? just (x, y)

jjatie replied:
If you look at the method implementation, it uses a centre point. The old parameter naming was misleading.

liuxuan30 commented on Source/Charts/Utils/ChartUtils.swift:
```diff
@@ -99,55 +99,65 @@ extension CGPoint
-        x: CGFloat,
-        y: CGFloat,
-        size: CGSize)
+    private static var _defaultValueFormatter: IValueFormatter = generateDefaultValueFormatter()
```
is this from another PR? Or should in this PR

jjatie replied:
This was just a minor change (ChartUtils.generateDefaultValueFormatter()` -> `generateDefaultValueFormatter()` because they're both static methods.) Nothing else. I can revert it if you want.

liuxuan30 commented:
you said at first:
>It is important to note that this is a breaking change because some of these methods were made public. However, it is extremely unlikely there are direct calls to them so I feel it is safe to merge them into the master. In 4.0.0 I would like to make all of these methods internal or at the very least, public, as our extension points don't need be at such a low level.

Is this better for 4.0 branch or master?

jjatie commented:
xcodeproj files have always been spaces though. At least since Xcode 3

jjatie commented:
(And no, I haven't changed my indentation settings. I leave them as default, but with "Include whitespace only lines" checked.)

jjatie commented:
Yea, I guess it should be on 4.0.0. I was hoping to get all the ChartUtils in master now, but it makes sense to do it in 4.0.0.

liuxuan30 commented:
So how do we handle the project file issue? Because on my side, all new created projects and Charts itself are showing tab instead of space, even my setting is using space:
![image](https://user-images.githubusercontent.com/4375169/34656085-88110598-f44f-11e7-8e44-92db6e35ff5c.png)
that's why I assume it's by Xcode default.

My concern is, it seems like a background scan on your side change those tabs/spaces. If Xcode update project file, it might replace spaces with tabs again.

Besides, have you tried other tools like using terminal, SourceTree, Github desktop, to check the diff?

There seems some options:
1. You fixed this weird scan
2. We reset --mix and don't commit project file when nothing really changed
3. Continue using yours

jjatie commented:
Yea, I have the same settings. I have no idea why it would have changed. As you can see, it's now all tabs again (I fixed it when I switched it over to 4.0.0 from master)

liuxuan30 commented:
And other people filing PR may submit this as well. You have been filed lots of PRs, so this is the first time we met this issue. That's why it's strange. Either you commit this at the very first PR, but it's in the middle..

jjatie commented:
I diffed it in BBEdit, yea.

liuxuan30 commented:
What?? You just pulled master and it's tab again? Really what's going on 😂

jjatie commented:
I really don't know haha. But this one is safe again. I'll manually change them to tabs if I have to in the other affected PRs, but I'll wait until you review them.

liuxuan30 commented:
sure. I do hope this never happens in the future. I will take a look at this PR again now and merge this, like in 30 mins

liuxuan30 commented on Source/Charts/Utils/ChartUtils.swift:
```diff
@@ -156,96 +166,96 @@ open class ChartUtils
             rect.origin.x = -knownTextSize.width * 0.5
             rect.origin.y = -knownTextSize.height * 0.5
-            
+
```
any chance this is something wrong with your setup? About deleting some spaces? You intend to do this or ?

jjatie replied:
I have remove white space in white space only lines turned on. No point in having a line of empty space.

liuxuan30 commented:
Just merged. You could move next one(just one lol) to 'In progress'

liuxuan30 commented:
I see we don't have more master branch PRs?

jjatie commented:
Yea, easier to maintain the 4.0 branch that way. They aren't explicit bug fixes or features so they can all wait until 4.0

liuxuan30 commented:
Then we can make it simple, you could drag several PRs to in progress, I will start from the top one.
And later maybe I can push master to pod then? bump as 3.0.5?

jjatie commented:
I think we can just bump to 3.0.5 now, unless there's an urgent bug fix lying around.
